src/Microsoft.AspNetCore.Razor.Design/build/netstandard2.0/Microsoft.AspNetCore.Razor.Design.targets

```diff
@@ -81,130 +92,71 @@
 
-    Notice there are plenty of Condition="'@(RazorGenerate)'!=''" to avoid running more expensive targets.
+    Most targets related to Razor code generation are defined there.
   -->
-  <Target Name="RazorResolveGenerateInputs">
-    <ItemGroup>
-      <RazorGenerate Include="@(Content)" Condition="'%(Content.Extension)'=='.cshtml'" />
-      <_RazorGenerateOutput Include="@(RazorGenerate->'$(RazorGenerateOutputPath)%(RelativeDir)%(Filename).cs')" />
-    </ItemGroup>
-  </Target>
+  <Import Project="Microsoft.AspNetCore.Razor.Design.CodeGeneration.targets" />
 
-  <!--
-    This target will only be called when we have some .cshtml files that are going to participate in code generation.
+  <!-- 
+    These are the targets that actually do compilation using CSC, separated for the main file for ease of maintenance.
 
-    This is part of the chain of targets that are called once we've actually committed to generating code.
+    RazorCoreCompile should be defined there.
   -->
-  <Target 
-    Name="_ResolveRazorCoreGenerateInputs" 
-    DependsOnTargets="RazorResolveGenerateInputs"
-    Condition="'@(RazorGenerate)'!=''">
+  <Import Project="Microsoft.AspNetCore.Razor.Design.Compilation.targets" />
+
+  <Target Name="PrepareForRazorGenerate" DependsOnTargets="$(PrepareForRazorGenerateDependsOn)">
+  </Target>
 
-    <PropertyGroup>
-      <_RazorGenerateHashFile>$(IntermediateOutputPath)$(MSBuildProjectName).RazorCoreGenerate.cache</_RazorGenerateHashFile>
-    </PropertyGroup>
+  <Target Name="RazorGenerate" DependsOnTargets="$(RazorGenerateDependsOn)">
+  </Target>
 
-    <Hash ItemsToHash="@(RazorGenerate)" Condition="'@(RazorGenerate)'!=''">
-      <Output TaskParameter="HashResult" PropertyName="_RazorGenerateHash" />
-    </Hash>
+  <Target Name="PrepareForRazorCompile" DependsOnTargets="$(PrepareForRazorCompileDependsOn)">
+  </Target>
 
-    <MakeDir
-      Directories="$(IntermediateOutputPath)"
-      Condition="!Exists('$(IntermediateOutputPath)')" />
+  <Target Name="RazorCompile" DependsOnTargets="$(RazorCompileDependsOn)">
+  </Target>
 
-    <WriteLinesToFile
-      Lines="$(_RazorGenerateHash)"
-      File="$(_RazorGenerateHashFile)"
-      Overwrite="True"
-      WriteOnlyWhenDifferent="True" />
+  <PropertyGroup Condition="'$(RazorCompileOnBuild)'=='true'">
+    <PrepareForRunDependsOn>RazorCompile;$(PrepareForRunDependsOn);_RazorCopyFilesToOutputDirectory</PrepareForRunDependsOn>
+    <GetCopyToOutputDirectoryItemsDependsOn>$(GetCopyToOutputDirectoryItemsDependsOn);_RazorGetCopyToOutputDirectoryItems</GetCopyToOutputDirectoryItemsDependsOn>
+  </PropertyGroup>
 
+  <!--
+    Gathers input source files for code generation. This is a separate target so that we can avoid
+    lots of work when there are no inputs for code generation.
+
+    Notice there are plenty of Condition="'@(RazorGenerate)'!=''" to avoid running more expensive targets.
+  -->
+  <Target Name="ResolveRazorGenerateInputs">
```

Comment:
This target is SDK-like so it belongs here. We won't have much reason to change this if as we evolve Razor.